airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumStateDecoratingIterator.java

```diff
@@ -0,0 +1,179 @@
+   * @param checkpointSeconds Seconds between syncs
+   * @param checkpointRecords Number of records between syncs
+   */
+  public DebeziumStateDecoratingIterator(final Iterator<ChangeEvent<String, String>> changeEventIterator,
+                                         final CdcStateHandler cdcStateHandler,
+                                         final CdcMetadataInjector cdcMetadataInjector,
+                                         final Instant emittedAt,
+                                         final AirbyteFileOffsetBackingStore offsetManager,
+                                         final boolean trackSchemaHistory,
+                                         final AirbyteSchemaHistoryStorage schemaHistoryManager,
+                                         final Integer checkpointSeconds,
+                                         final Integer checkpointRecords) {
+    this.changeEventIterator = changeEventIterator;
+    this.cdcStateHandler = cdcStateHandler;
+    this.cdcMetadataInjector = cdcMetadataInjector;
+    this.emittedAt = emittedAt;
+    this.offsetManager = offsetManager;
+    this.trackSchemaHistory = trackSchemaHistory;
+    this.schemaHistoryManager = schemaHistoryManager;
+
+    this.syncCheckpointDuration = Duration.ofSeconds(checkpointSeconds);
+    this.syncCheckpointRecords = checkpointRecords;
+    this.previousCheckpointOffset = (HashMap<String, String>) offsetManager.read();
+    resetCheckpointValues();
+  }
+
+  /**
+   * Computes the next record retrieved from Source stream. Emits state messages as checkpoints based
+   * on number of records or time lapsed.
+   *
+   * <p>
+   * If this method throws an exception, it will propagate outward to the {@code hasNext} or
+   * {@code next} invocation that invoked this method. Any further attempts to use the iterator will
+   * result in an {@link IllegalStateException}.
+   * </p>
+   *
+   * @return {@link AirbyteStateMessage} containing CDC data or state checkpoint message.
+   */
+  @Override
+  protected AirbyteMessage computeNext() {
+    if (isSyncFinished) {
+      return endOfData();
+    }
+
+    if (sendCheckpointMessage) {
+      AirbyteMessage stateMessage = createStateMessage(checkpointOffset);
+      previousCheckpointOffset.clear();
+      previousCheckpointOffset.putAll(checkpointOffset);
+      resetCheckpointValues();
+      return stateMessage;
+    }
+
+    if (changeEventIterator.hasNext()) {
+      try {
+        final ChangeEvent<String, String> event = changeEventIterator.next();
+        recordsLastSync++;
+
+        if (checkpointOffset.size() == 0 &&
+            (recordsLastSync >= syncCheckpointRecords ||
+                Duration.between(dateTimeLastSync, OffsetDateTime.now()).compareTo(syncCheckpointDuration) > 0)) {
+          checkpointOffset.putAll(offsetManager.read());
+          if (!previousCheckpointOffset.isEmpty() &&
+              cdcStateHandler.isSameOffset(checkpointOffset, previousCheckpointOffset)) {
```

Comment:
This can happen if 15 minutes passed with no record?

Reply:
What do you mean? If there are no records since we started the sync, both `previousCheckpointOffset` and `checkpointOffset` will be at the same offset (in both cases we are assigning the value `offsetManager.read()`).

Reply:
You answer what I meant I think. If no records since last sync then `cdcStateHandler.isSameOffset(checkpointOffset, previousCheckpointOffset)` is `true`.

What I'm also thinking, in case of a very big db where almost no records relate to the stream we sync, there might pass 15 minutes of reading CDC logs without any single relevant ChangeEvent coming (all from tables we do not sync)

Reply:
In that case, it will be trying to get the offset from the records included in the stream to sync... otherwise it will just set the final offset, but I got your point. BTW, if that's the case... the client should re-think how they organise their data 🤣 

I'm not sure if we can play with the heartbeat one.. 🤔